MI5-38 Kreiran endpoint za registraciju

kmilos-foi opened the pull request:
Uspostavljen je funkcionalan endpoint za proces registracije. Implementirana je validacija podataka novog korisnika i hashiranje lozinke prije nego što bude zapisana u bazu podataka. Prije registracije dodatno se provjerava valjanost poslanog JWT-a. Ako JWT nije valjan, registracija korisnika neće se izvršiti. Također, u slučaju da bilo koji dio korisničkih podataka ne prođe validaciju, korisnik neće biti spremljen u bazu podataka.

kmilos-foi committed:
MI5-38 Kreiran UserService

kmilos-foi committed:
MI5-38 Kreiran UserController

kmilos-foi committed:
MI5-38 Kreiran Email validator

Klasa sadrži statičku metodu za provjeru formata emaila pomoću regexa

kmilos-foi committed:
MI5-38 Dodane existsByEmail i existsByUsername

Metode u repozitoriju koje provjeravaju postoji li korisnik s prosljeđenim emailom ili korisničkim imenom.

kmilos-foi committed:
MI5-38 Stvoren UserDataValidator

EmailValidator prebačen u UserDataValidator

kmilos-foi committed:
MI5-38 Stvoren UserService

kmilos-foi committed:
MI5-38 Stvoren interface Validator

kmilos-foi committed:
MI5-38 Refaktoriran UserDataValidator

kmilos-foi committed:
MI5-38 Dodana metoda za verifikaciju jwt-a

kmilos-foi committed:
MI5-38 Stvoren JwtService

kmilos-foi committed:
MI5-38 Dodana provjera jwt-a u UserControlleru

UserController koristi JwtService koji je zadužen za verifikaciju dobivenog tokena

kmilos-foi committed:
MI5-38 Dodana metoda za dohvat uloge

Potrebno je dohvatiti ulogu iz JWT-a kako bi se moglo provjeriti je li korisnikova uloga autorizirana za određenu akciju ili pristup resursu.

kmilos-foi committed:
MI5-38 Dodano hashiranje lozinke

Lozinka se hashira prije spremanja u bazu podataka

kmilos-foi committed:
MI5-38 Dodana provjera uloge

Metoda koja provjerava je li uloga admin.

kmilos-foi committed:
MI5-38 Ažuriran UserController

kmilos-foi committed:
MI5-38 Dodan AuthorizationException

kmilos-foi committed:
MI5-38 Refaktoriran UserDataValidator

kmilos-foi committed:
MI5-38 Stvoren ValidationException

kmilos-foi committed:
MI5-38 Stvoren JwtException

kmilos-foi committed:
MI5-38 Ažuriran UserController

kmilos-foi committed:
MI5-38 Stvoren RoleRepository

kmilos-foi committed:
MI5-38 Ažuriran UserDataValidator

Kod je refaktoriran.

kmilos-foi committed:
MI5-38 Dodana metoda dodjelu uloge

Postavljanje cijelog objekta UserRole na temelju naziva UserRole-a.

kmilos-foi committed:
MI5-38 Ažuriran UserController

Dio koda za provjeru je li uloga administrator premješen u try-catch blok.

NinoLoncar commented on src/main/java/foi/air/szokpt/accountmng/controllers/UserController.java:
Mislim da bi trebalo poraditi na refaktorizaciji ovog kontrolera. Pogledaj ExceptionHandler ili RestControllerAdvice anotacije za upravljanje iznimkama (https://medium.com/@seonggil/handling-exceptions-with-restcontrolleradvice-exceptionhandler-e7c95216da8d).

bkadezabe20 commented on src/main/java/foi/air/szokpt/accountmng/services/UserService.java:
```diff
@@ -0,0 +1,43 @@
+        userDataValidator.validateData(user);
+        assignRoleToUser(user);
+        user.setPassword(hasher.hashText(user.getPassword()));
+        userRepository.save(user);
```
Tu bi trebao biti try catch implementirani

```suggestion
        try {
               userRepository.save(user);
        } catch (Exception e) {
               throw new ValidationException("Failed to register user.");
         }
```

kmilos-foi committed:
MI5-38 Stvoren GlobalExceptionHandler

kmilos-foi committed:
MI5-38 Refaktorirana logika oko registracije

Sva logika koja je bila sadržana u UserControlleru sada je prebačena na UserService. Više nema potrebe za JwtServiceom. U JwtUtil je prebačena metoda extractToken.

kmilos-foi committed:
MI5-38 Riješen problem provjere emaila

Email adresa se pretvara u mala slova prije provjere njezinog postojanja u bazi podataka, a zatim se ponovno konvertira u mala slova u UserService klasi. Time se riješio problem kada se email ne pošalje na server (u tom slučaju bi se email pretvarao u mala slova, a da email uopće ne postoji, što rezultira nepoznatim problemom).
